crml/nft/src/lib.rs

```diff
@@ -0,0 +1,651 @@
+		AuctionSold(CollectionId, TokenId, AssetId, Balance, AccountId),
+		/// An auction has closed without selling (collection, token, reason)
+		AuctionClosed(CollectionId, TokenId, Reason),
+		/// A new highest bid was placed (collection, token, amount)
+		Bid(CollectionId, TokenId, Balance),
+	}
+);
+
+decl_error! {
+	/// Error for the staking module.
+	pub enum Error for Module<T: Trait> {
+		/// A collection with the same ID already exists
+		CollectionIdExists,
+		/// Given collection ID is not valid utf-8
+		CollectionIdInvalid,
+		/// No more Ids are available, they've been exhausted
+		NoAvailableIds,
+		/// Max tokens issued
+		MaxTokensIssued,
+		/// Too many attributes in the provided schema or data
+		SchemaMaxAttributes,
+		/// Provided attributes do not match the collection schema
+		SchemaMismatch,
+		/// Provided attribute is not in the collection schema
+		UnknownAttribute,
+		/// The provided attributes or schema cannot be empty
+		SchemaEmpty,
+		/// The schema contains an invalid type
+		SchemaInvalid,
+		/// The schema contains a duplicate attribute name
+		SchenmaDuplicateAttribute,
+		/// Given attirbute value is larger than the configured max.
+		MaxAttributeLength,
+		/// origin does not have permission for the operation
+		NoPermission,
+		/// The NFT collection does not exist
+		NoCollection,
+		/// The NFT does not exist
+		NoToken,
+		/// The NFT is not listed for a direct sale
+		NotForDirectSale,
+		/// The NFT is not listed for auction sale
+		NotForAuction,
+		/// Cannot operate on a listed NFT
+		TokenListingProtection,
+		/// Internal error during payment
+		InternalPayment,
+		/// Total royalties would exceed 100% of sale
+		RoyaltiesOvercommitment,
+		/// Auction bid was lower than reserve or current highest bid
+		BidTooLow
+	}
+}
+
+decl_storage! {
+	trait Store for Module<T: Trait> as Nft {
+		/// Map from collection to owner address
+		pub CollectionOwner get(fn collection_owner): map hasher(blake2_128_concat) CollectionId => Option<T::AccountId>;
+		/// Map from collection to its schema definition
+		pub CollectionSchema get(fn collection_schema): map hasher(blake2_128_concat) CollectionId => Option<NFTSchema>;
+		/// Map from collection to it's defacto royalty scheme
+		pub CollectionRoyalties get(fn collection_royalties): map hasher(blake2_128_concat) CollectionId => Option<RoyaltiesSchedule<T::AccountId>>;
+		/// Map from a token to it's royalty scheme
+		pub TokenRoyalties get(fn token_royalties): double_map hasher(blake2_128_concat) CollectionId, hasher(twox_64_concat) T::TokenId => Option<RoyaltiesSchedule<T::AccountId>>;
+		/// Map from (collection, token) to it's attributes (as defined by schema)
+		pub Tokens get(fn tokens): double_map hasher(blake2_128_concat) CollectionId, hasher(twox_64_concat) T::TokenId => Vec<NFTAttributeValue>;
+		/// The next available token Id for an NFT collection
+		pub NextTokenId get(fn next_token_id): map hasher(twox_64_concat) CollectionId => T::TokenId;
```

Comment:
Can we use the `last` function of the iterator to the Tokens storage and avoid this storage entirely.

Reply:
No because if a token is burnt it will be removed from the tokens map.
Should use `decode_len` rather than iterator for getting item count

Reply:
Even when a token is burnt, the last token should have the token id from which we can generate the next one.

Reply:
what if the last token is burnt?

Reply:
That's a good question. Would it be a problem to reuse the token id in that case? I'm thinking we'd burn tokens to free their resources. There should be a good reason why a token id should remain reserved for a burnt token.

Reply:
it's much simpler this way, not much gain to optimise this